cpp/src/arrow/chunked_array.h

```diff
@@ -263,8 +263,10 @@ Status ApplyBinaryChunked(const ChunkedArray& left, const ChunkedArray& right,
                           Action&& action) {
   MultipleChunkIterator iterator(left, right);
   std::shared_ptr<Array> left_piece, right_piece;
+  int64_t pos = iterator.position();
   while (iterator.Next(&left_piece, &right_piece)) {
-    ARROW_RETURN_NOT_OK(action(*left_piece, *right_piece, iterator.position()));
+    ARROW_RETURN_NOT_OK(action(*left_piece, *right_piece, pos));
+    pos = iterator.position();
```

Comment:
Why do we need this change?

Reply:
The pre-PR code reads `iterator.position()` after it is advanced, so the action gets a 1-based column index, which is incorrect per [the documentation comments](https://github.com/apache/arrow/blob/febd0ff144cfb8b2baffb1cb0be57ca40dc7cc77/cpp/src/arrow/chunked_array.h#L239-L260). In the investigation here, this index ended up in the messages and got me confused until I discovered this.

Reply:
Ah, I see.  This does appear to be more or less unused outside of test reporting so perhaps it just wasn't noticed.  `MultiChunkIterator::position` is very unintuitive.  However, given its been around a few years, we should probably leave it alone.  Thanks for this fix.